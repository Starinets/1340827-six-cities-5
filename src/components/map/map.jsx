import React from 'react';
import * as Type from '../../types';

import Leaflet from 'leaflet';
import 'leaflet/dist/leaflet.css';

import {
  START_MAP_POSITION,
  START_MAP_ZOOM,
  MAP_LAYER,
  MAP_ATTRIBUTION,
  Icon,
  ActiveIcon
} from '../../constants';

class Map extends React.PureComponent {

  constructor(props) {
    super(props);

    this.offers = props.offers;
    this.hoveredOffer = props.hoveredOffer;
    this.mapPlace = props.mapPlace;

    this.markers = [];
    this.map = undefined;
  }

  _renderMarkers() {

    const icon = Leaflet.icon(Icon);
    const activeIcon = Leaflet.icon(ActiveIcon);

    this.offers.forEach((offer) => {

      const offerCords = [offer.latitude, offer.longitude];
      const currentIcon = offer === this.hoveredOffer ? activeIcon : icon;
      const marker = Leaflet.marker(offerCords, {icon: currentIcon});

      marker.addTo(this.map);
      this.markers.push(marker);
    });
  }

  componentDidMount() {

    this.map = Leaflet.map(`map`, {
      center: START_MAP_POSITION,
      zoom: START_MAP_ZOOM,
      zoomControl: false,
      marker: true
    });

    this.map.setView(START_MAP_POSITION, START_MAP_ZOOM);

    Leaflet.tileLayer(MAP_LAYER,
        {
          attribution: MAP_ATTRIBUTION
        }).addTo(this.map);

    this._renderMarkers();
  }

  componentWillUpdate(props) {

    this.markers.forEach((marker) => {
      this.map.removeLayer(marker);
    });

    this.markers = [];
    this.offers = props.offers;
    this.hoveredOffer = props.hoveredOffer;
    this._renderMarkers();
  }

  render() {
    return (
      <section className={ `${ this.mapPlace }__map map` } id="map"></section>
    );
  }
}

Map.propTypes = {
  offers: Type.OFFERS.isRequired,
  mapPlace: Type.MAP_PLACE.isRequired,
  hoveredOffer: Type.OFFER
};

export default Map;
