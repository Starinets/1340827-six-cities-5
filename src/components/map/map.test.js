import React from 'react';
import renderer from 'react-test-renderer';
import {Map} from './map';
import {
  Offers,
  MapPlace,
  offer
} from '../../test-data/test-data';

describe(`Component rendered correctly -> Map`, () => {
  it(`on the Main page`, () => {
    const div = document.createElement(`div`);
    div.id = `map`;
    document.body.appendChild(div);

    const tree = renderer.create(
        <Map
          offers={ Offers }
          mapPlace={ MapPlace.CITIES }
          hoveredOffer={ offer }
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`on the Offer page`, () => {
    document.getElementById(`map`).remove();
    const div = document.createElement(`div`);
    div.id = `map`;
    document.body.appendChild(div);

    const tree = renderer.create(
        <Map
          offers={ Offers }
          mapPlace={ MapPlace.OFFER }
          hoveredOffer={ offer }
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
