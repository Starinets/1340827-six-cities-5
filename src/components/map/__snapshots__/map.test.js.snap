// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component rendered correctly -> Map on the Main page 1`] = `
<section
  className="cities__map map"
  id="map"
/>
`;

exports[`Component rendered correctly -> Map on the Offer page 1`] = `
<section
  className="property__map map"
  id="map"
/>
`;
